geoutils/georaster.py

```diff
@@ -822,10 +913,179 @@ def format_value(value):
+        if return_window:
             return (value, win)
         else:
             return value
-    
+
+    def xxyy(self, offset='corner', grid=True):
+        """
+        Get x,y coordinates of all pixels in the raster.
+
+        :param offset: coordinate type. If 'corner', returns corner coordinates of pixels.
+            If 'center', returns center coordinates. Default is corner.
+        :param grid: Return gridded coordinates. Default is True.
+        :type offset: str
+        :type grid: bool
+        :returns x,y: numpy arrays corresponding to the x,y coordinates of each pixel.
+        """
+        assert offset in ['corner', 'center'], "ctype is not one of 'corner', 'center': {}".format(offset)
+
+        xmin, ymin, xmax, ymax = self.ds.bounds
+        dx = list(self.ds.transform)[0]
+        dy = list(self.ds.transform)[4]
+        npix_x = self.ds.width
+        npix_y = self.ds.height
+
+        # transform = Affine.from_gdal(self.ds.transform)
+        # nx, ny = self.ds.width, self.ds.height
+        # xx, yy = np.meshgrid(np.arange(nx) + 0.5, np.arange(ny) + 0.5) * transform
+
+        # return xx, yy
+
+        if dx < 0:
+            xx = np.linspace(xmax, xmin, npix_x + 1)
+        else:
+            xx = np.linspace(xmin, xmax, npix_x + 1)
+
+        if dy < 0:
+            yy = np.linspace(ymax, ymin, npix_y + 1)
+        else:
+            yy = np.linspace(ymin, ymax, npix_y + 1)
+
+        if offset == 'center':
+            xx += dx / 2  # shift by half a pixel
+            yy += dy / 2
+        if grid:
+            return np.meshgrid(xx[:-1], yy[:-1])  # drop the last element
+        else:
+            return xx[:-1], yy[:-1]
+
+    def xy2ij(self,x,y):
+        """
+        Return row, column indices for a given x,y coordinate pair.
+
+        :param x: x coordinates
+        :type x: array-like
+        :param y: y coordinates
+        :type y: array-like
+
+        :returns i, j: indices of x,y in the image.
+        :rtype i, j: array-like
+
+        """
+        i, j = self.ds.index(x,y)
+
+        return i, j
+
+    def ij2xy(self,i,j,offset='center'):
+
+        """
+        Return x,y coordinates for a given row, column index pair.
+
+        :param i: row (i) index of pixel.
+        :type i: array-like
+        :param j: column (j) index of pixel.
+        :type j: array-like
+        :param offset: return coordinates as "corner" or "center" of pixel
+        :param offset: str
+
+        :returns x, y: x,y coordinates of i,j in reference system.
+        """
+
+        x,y = self.ds.xy(i,j,offset=offset)
+
+        return x, y
+
+    def outside_image(self, xi,yj, index=True):
+        """
+        #TODO: calculate matricially for all points instead of doing for only one?
+        Check whether a given point falls outside of the raster.
+
+        :param xi: Indices (or coordinates) of x direction to check.
+        :param yj: Indices (or coordinates) of y direction to check.
```

Comment:
Ok, I see you do it again. Is it not more standard to have the type just after the param, rather than all the types at the end?

Reply:
It is indeed more standard to do: param type param type ([cf doc](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html))
This was copied from other code and I simply forgot to change (so it's basically Bob's fault @iamdonovan )

Reply:
Of course, it's Bob's fault!